vod/media_set_parser.c

```diff
@@ -783,6 +802,121 @@ media_set_sequence_id_exists(request_params_t* request_params, vod_str_t* id)
+	{
+		vod_log_error(VOD_LOG_ERR, request_context->log, 0,
+			"media_set_parse_closed_captions: invalid closed caption type %d expected object", array->type);
+		return VOD_BAD_MAPPING;
+	}
+	if (array->count > MAX_CLOSED_CAPTIONS)
+	{
+		vod_log_error(VOD_LOG_ERR, request_context->log, 0,
+		"media_set_parse_closed_captions: invalid number of elements in the closed captions array %uz", array->count);
+		return VOD_BAD_MAPPING;
+	}
+	media_set->closed_captions_count = array->count;
+
+	cur_output = vod_alloc(request_context->pool, sizeof(cur_output[0]) * array->count);
+	if (cur_output == NULL)
+	{
+		vod_log_debug0(VOD_LOG_DEBUG_LEVEL, request_context->log, 0,
+			"media_set_parse_closed_captions: vod_alloc failed");
+		return VOD_ALLOC_FAILED;
+	}
+
+	media_set->closed_captions = cur_output;
+
+	part = &array->part;
+	for (cur_pos = part->first; ; cur_pos++)
+	{
+		if ((void*)cur_pos >= part->last)
+		{
+			if (part->next == NULL)
+			{
+				break;
+			}
+
+			part = part->next;
+			cur_pos = part->first;
+		}
+
+		vod_memzero(params, sizeof(params));
+
+		vod_json_get_object_values(cur_pos, &media_closed_captions_hash, params);
+
+		if (params[MEDIA_CLOSED_CAPTIONS_PARAM_ID] == NULL)
+		{
+			vod_log_error(VOD_LOG_ERR, request_context->log, 0,
+				"media_set_parse_closed_captions: missing id in closed captions object");
+			return VOD_BAD_MAPPING;
+		}
+		if (params[MEDIA_CLOSED_CAPTIONS_PARAM_LANGUAGE] == NULL)
+		{
+			vod_log_error(VOD_LOG_ERR, request_context->log, 0,
+				"media_set_parse_closed_captions: missing language in closed captions object");
+			return VOD_BAD_MAPPING;
+		}
+		if (params[MEDIA_CLOSED_CAPTIONS_PARAM_LABEL] == NULL)
+		{
+			vod_log_error(VOD_LOG_ERR, request_context->log, 0,
+				"media_set_parse_closed_captions: missing label in closed captions object");
+			return VOD_BAD_MAPPING;
+		}
+
+		rc = media_set_parse_language(request_context, params[MEDIA_CLOSED_CAPTIONS_PARAM_LANGUAGE], &cur_output->language);
+		if (rc != VOD_OK)
+    {
+      return rc;
+    }
+
+		id = &params[MEDIA_CLOSED_CAPTIONS_PARAM_ID]->v.str;
+		cur_output->id.data = vod_alloc(request_context->pool, id->len + 1);
+		if (cur_output->id.data == NULL)
+		{
+			vod_log_debug0(VOD_LOG_DEBUG_LEVEL, request_context->log, 0,
+			"media_set_parse_closed_captions: vod_alloc failed");
+			return VOD_ALLOC_FAILED;
+		}
+
+		vod_memcpy(cur_output->id.data, id->data, id->len);
+		cur_output->id.data[id->len] = '\0';
+		cur_output->id.len = id->len;
+
+		label = &params[MEDIA_CLOSED_CAPTIONS_PARAM_LABEL]->v.str;
+		cur_output->label.data = vod_alloc(request_context->pool, label->len + 1);
+		if (cur_output->label.data == NULL)
+		{
+			vod_log_debug0(VOD_LOG_DEBUG_LEVEL, request_context->log, 0,
+			"media_set_parse_closed_captions: vod_alloc failed");
+			return VOD_ALLOC_FAILED;
+		}
+
+		vod_memcpy(cur_output->label.data, label->data, label->len);
+		cur_output->label.data[label->len] = '\0';
+		cur_output->label.len = label->len;
```

Comment:
better use media_set_parse_null_term_string twice here, it will also decode the string (the label may contain some escaped chars)